File: restore_ip_addresses/restore_ip_addresses.py
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        # the most straightforward way to reason about this problem is to think of s as some kind of queue or stack where we look at the indexes in three digit increments and see if they fullfill specific invalidation rules. 
            # is the three digit increments an actual number?
            # is the number between 0 and 255

        # if they pass these validation principles than we can add them to the state array.

        # base case will be if the state array matches a length of four.

        # recursive case will be if the state array is less than the length of four.

        # create a results array
        # create a recursive function that will take state, s, results as arguments

        # return results

        results = []
        self.recursive_helper([], s, results)
        return results

    def recursive_helper(self, state: List[str], s: str, results: List[List[str]]):
        # base case if the state variable has a length equal to 4 than add it to the results array. Make sure that the state variable is converted to a string with a . delimiter.

        # recursive case
        # if the state variable has a length variable less than 4.
            # create a for loop that will loop through three digit spots of the s variable 
                # create a conditional check to see if s[:i] are valid ip address integers.
                    # if true add s[:i] to the state array
                    # recursively call recursive_helper with updated state array, s[:i+1], and results array
                    # deselect s[:i] in the state array

        if len(state) == 4 and len(s) == 0:
            state_str = ".".join(state[:])
            results.append(state_str)
            return 
        elif len(state) < 4 and len(s) > 0:
            for i in range(1, 4):
                if i > len(s):
                    break
                
                if self.ip_digit_validator(s[0:i]):
                    state.append(s[0:i])
                    self.recursive_helper(state, s[i:], results)
                    state.pop()

    def ip_digit_validator(self, digit: str) -> bool:
        if digit[0] == "0" and len(digit) > 1:
            return False
        if int(digit) <= 255:
            return True
        
        return False

File: restore_ip_addresses/restore_ip_addresses.go
package restoreipaddresses

import (
	"strconv"
	"strings"
)

func restoreIpAddresses(s string) []string {
	// first I will need to create a results array that will carry all the conbinations of valid ip addresses

	// create a recursive function that will take a state argument, s, and results

	// return results

	var results []string
	var state []string
	recursion_helper(state, s, &results)
	return results
}

func recursion_helper(state []string, s string, results *[]string) {

	if len(state) == 4 && len(s) == 0 {
		state_str := strings.Join(state[:], ".")
		*results = append(*results, state_str)
		return
	} else if len(state) < 4 && len(s) > 0 {
		for i := 1; i < 4; i++ {
			if i > len(s) {
				break
			}

			if checkIpDigit(s[0:i]) {
				state = append(state, s[0:i])
				recursion_helper(state, s[i:], results)
				state = state[:len(state)-1]
			}
		}
	}
}

func checkIpDigit(digit string) bool {
	if string(digit[0]) == "0" && len(digit) > 1 {
		return false
	}

	if i, _ := strconv.Atoi(digit); i <= 255 {
		return true
	}

	return false
}
